File: 프로그래머스/lv2/87946. 피로도/피로도.kt
class Solution {
    var answer = 0
    var visited = BooleanArray(0)
    fun solution(k: Int, dungeons: Array<IntArray>): Int {
        visited = BooleanArray(dungeons.size)
        
        dfs(0, k, dungeons)
        return answer
    }
    
    fun dfs(depth:Int, p:Int, dungeons: Array<IntArray>){
        for(i in 0 until dungeons.size){
            if(!visited[i] && dungeons[i][0] <= p){
                visited[i] = true
                dfs(depth+1, p-dungeons[i][1], dungeons)
                  visited[i] = false
            }
            
            answer = Math.max(answer, depth)
        }
    }
    
}

File: 프로그래머스/2/87946. 피로도/피로도.js
let max_result = 0;

function solution(k, dungeons) {
    const visited = Array(dungeons.length).fill(false);
    dfs(k, dungeons, visited, 0);
    return max_result;
}

function dfs(k, dungeons, visited, result) {
    max_result = Math.max(max_result, result);

    for (let i = 0; i < dungeons.length; i++) {
        const [min_required, cost] = dungeons[i];
        if (!visited[i] && k >= min_required) {
            visited[i] = true;
            dfs(k - cost, dungeons, visited, result + 1);
            visited[i] = false;
        }
    }
}
